feat(ui): add TPS metric to token usage display

Mahadevan2005 opened the pull request:
## Summary
Adds a TPS (tokens per second) metric to the `TokenUsageDisplay` component to show token generation speed during LLM responses.

## Related Issue
Closes #72

## Changes Made
- Added `useMemo` hook to calculate TPS from `completion_tokens`, `ttr`, and `ttft`
- Formula: `completion_tokens / ((ttr - ttft) / 1000)` tokens per second
- Added TPS display with `MdSpeed` icon formatted as "X t/s"
- Added tooltip: "Generation Speed: X tokens/sec"
- Used `hasTPS` guard for consistent render pattern (matches `hasTokenUsage`, `hasTTFT`, `hasTTR`)
- Positioned after TTR metric in the usage bar
- Handles edge cases: null values, invalid calculations, division by zero

## Type of Change
- [x] New feature (non-breaking change which adds functionality)

## Testing
- [x] TypeScript compilation passes (`npm run build`)
- [x] Linting passes - no new errors introduced
- [x] Build succeeds in production mode
- [x] Code follows existing patterns and conventions
- [x] Follows Skyflo.ai coding standards (TypeScript, Airbnb style)

## Verification
```
✓ Compiled successfully
✓ Checking validity of types
✓ No linting errors in TokenUsageDisplay.tsx
✓ Build completed successfully
```

## Notes
- No new linting errors introduced - TokenUsageDisplay.tsx passes all checks
- Pre-existing linting errors in other files remain unchanged
- Implementation follows existing component patterns (`has*` naming convention)

Mahadevan2005 commented:
@KaranJagtiani the changes are done! 

- Squashed all commits into one
- No lock file changes
- Only 2 code files modified

PR is now clean and ready.

KaranJagtiani commented on ui/src/components/chat/ChatInterface.tsx:
The `lastTTRUpdateRef` needs to be reset here as well.

```suggestion
        lastTTRUpdateRef.current = 0;
        requestStartTimeRef.current = Date.now();
```

Mahadevan2005 commented:
@KaranJagtiani I've implemented issue #75! Here's what was done:

### Backend Changes:
- Added `duration_ms` (int) to completed SSE event payload in `graph.py`
- Updated `record_ttr()` in `agent.py` to prefer `duration_ms` over `duration * 1000`
- Maintained backwards compatibility with existing `duration` field

### Frontend Changes:
- Added `duration_ms?: number` to `CompletedEvent` type
- Updated `onComplete` callback to accept optional `duration_ms` parameter  
- Parse and propagate server duration from SSE events
- Prefer server-measured TTR over client calculation
- Keep client timing as fallback for older engines
- Added missing `lastTTRUpdateRef` resets

### Result:
- TPS now uses accurate server timing (no network latency)
- TTR remains consistent when reloading conversations
- No breaking changes to SSE contract
- Fully backwards compatible

KaranJagtiani commented on ui/src/components/chat/ChatInterface.tsx:
```diff
@@ -457,9 +459,19 @@ export function ChatInterface({ conversationId }: ChatInterfaceProps) {
-        if (requestStartTimeRef.current) {
-          const ttr = Date.now() - requestStartTimeRef.current;
+      onComplete: (duration_ms?: number) => {
+        // Prefer server-measured duration_ms, fallback to client timing
```
Remove redundant comments.

KaranJagtiani commented on engine/src/api/agent/graph.py:
We need to emit the `duration` / `duration_ms` for stopped runs too.

```suggestion
            except StopRequested:
                end_time = time.time()
                start_time = get_state_value(initial_state, "start_time", end_time)
                duration = end_time - start_time
                duration_ms = int(duration * 1000)

                if self.event_callback:
                    await self.event_callback(
                        {
                            "type": "completed",
                            "status": "stopped",
                            "run_id": get_state_value(initial_state, "run_id"),
                            "duration": duration,
                            "duration_ms": duration_ms,
                        }
                    )
```

KaranJagtiani commented on ui/src/components/chat/ChatInterface.tsx:
The same block never resets requestStartTimeRef or lastTTRUpdateRef before firing the next approval stream. Every approval after the first therefore measures elapsed time starting from the very first tool. While the server now emits duration_ms at completion, all streamed TPS updates (and any fallback path when duration_ms is absent) are computed off this stale timestamp, so the denominator grows across the entire batch and misstates generation speed. Reset the start time and throttle ref before invoking startApprovalStream.

Mahadevan2005 commented:
@KaranJagtiani All bulk approval and aggregate TPS issues fixed!

**Changes Made:**

### Bulk Approval Fixes:
- Added `resetLiveUsage()` before each approval in bulk continuation
- Added `lastTTRUpdateRef.current = 0` before each approval
- Added `requestStartTimeRef.current = Date.now()` before each approval
- Prevents metric accumulation across sequential approvals

### Aggregate TPS Fixes:
- Added `total_generation_ms` to TokenUsage type
- Accumulate generation time in `accumulateUsage()` function
- Updated TPS calculation to prefer `total_generation_ms` for aggregates
- Keep per-message TPS using `(ttr - ttft)` for individual displays

**Testing Logic:**

**Single Turn:**
- Uses `(ttr - ttft)` → Per-message TPS

**Multi-Turn Footer:**
- Turn 1: 10 tokens, 1000ms generation
- Turn 2: 20 tokens, 1000ms generation
- Turn 3: 15 tokens, 1000ms generation
- Total: 45 tokens / 3000ms = 15 t/s

**Bulk Approvals:**
- Approval 1: Fresh metrics from 0
- Approval 2: Fresh metrics from 0 (resets before starting)
- Approval 3: Fresh metrics from 0 (resets before starting)

TPS feature is now ready for all workflows.

KaranJagtiani commented on ui/src/components/chat/ChatInterface.tsx:
```diff
@@ -594,14 +618,29 @@ export function ChatInterface({ conversationId }: ChatInterfaceProps) {
+            ? now - requestStartTimeRef.current
+            : undefined;
+
+        // Throttle TTR updates so we don't re-render too often while streaming
```
```suggestion
        // Throttle TTR updates to prevent excessive re-renders during streaming
```

Mahadevan2005 commented:
@KaranJagtiani Done! 

- Updated comment wording as requested
- Squashed all commits into one

Ready to merge!

KaranJagtiani reviewed (approved):
Great work @Mahadevan2005 👏 Closing two issues in a single PR is well done.

Mahadevan2005 commented:
@KaranJagtiani Thank you so much for the thorough reviews and guidance throughout this PR! 🙏

This was an incredible learning experience - from implementing the initial TPS metric to expanding the scope to fix the server-measured TTR architecture. Your detailed feedback helped me understand not just what to change, but why it matters.

Special thanks for:
- Catching the bulk approval metric accumulation issue
- Guiding the aggregate TPS calculation fix
- The patience through 15+ review cycles
- Teaching me about backwards compatibility and proper API design

I learned a ton about, real-time systems, and professional code review processes. Looking forward to contributing more to Skyflo.ai! 🚀

Thanks again for the mentorship!
